File: 04_Awari/perl/awari.pl
#!/usr/bin/env perl
use v5.24;
use warnings;
use experimental 'signatures';
no warnings 'experimental::signatures';
use List::Util 'none';

# our board will be represented with an array of 14 slots, from 0 to 13.
# Positions 6 and 13 represent the "home pit" for the human and the
# computer, respectively.
use constant PLAYER_HOME => 6;
use constant COMPUTER_HOME => 13;

use constant FIRST => 0;
use constant AGAIN => 1;

exit main(@ARGV);

sub main {
   $|++; # disable buffering on standard output, every print will be
         # done immediately

   welcome(); # startup message

   # this array will keep track of computer-side failures, defined as
   # "the computer did not win". Whenever the computer loses or draws, the
   # specific sequence of moves will be saved and then used to drive
   # the search for a (hopefully) optimal move.
   my $failures = [];
   while ('enjoying') {

      # a new game starts, let's reset the board to the initial condition
      my $board = [ (3) x 6, 0, (3) x 6, 0 ];

      # this string will keep track of all moves performed
      my $moves = '/';

      # the human player starts
      my $turn = 'player';

      say "\n";
      print_board($board);

      while (not is_game_over($board)) {

         my $move; # this will collect the move in this turn

         if ($turn eq 'player') { # "first" move for player

            # player_move(...) does the move selected by the player,
            # returning both the selected move as well as the pit id
            # where the last seed landed
            ($move, my $landing) = player_move($board);

            # if we landed on the Player's Home Pit we get another move
            $turn = $landing == PLAYER_HOME ? 'player-again' : 'computer';
         }
         elsif ($turn eq 'player-again') { # "second" move for player

            # here we call player_move making it clear that it's the
            # second move, to get the right prompt eventually. We only
            # care for the $move as the result, so we ignore the other.
            ($move) = player_move($board, AGAIN);
            $turn = 'computer';
         }
         else {

            # the computer_move(...) function analyzes the $board as well
            # as adapting the strategy based on past "failures" (i.e.
            # matches where the computer did not win). For this it's
            # important to pass the log of these failures, as well as the
            # full record of moves in this specific match.
            ($move, my $landing) = computer_move($board, $failures, $moves);
            print "\nMY MOVE IS ", $move - 6;

            # do the second move in the turn if conditions apply
            if ($landing == COMPUTER_HOME && ! is_game_over($board)) {

               # save the first move before doing the second one!
               $moves .= "$move/";

               my ($move) = computer_move($board, $failures, $moves);
               print ',', $move - 6;
            }
            $turn = 'player';
         }

         # append the last selected move by either party, to track this
         # specific match (useful for computer's AI and ML)
         $moves .= "$move/";
         print_board($board);
      }

      # assess_victory() returns the difference between player's and
      # computer's seeds, so a negative value is a win for the computer.
      my $computer_won = assess_victory($board) < 0;

      # if this last match was a "failure" (read: not a win for the
      # computer), then record it for future memory.
      push $failures->@*, $moves unless $computer_won;
   }

   return 0;
}

# calculate the difference between the two home pits. Negative values mean
# that the computer won, 0 is a draw, positive values is a player's win.
# The difference is also returned back, in case of need.
sub assess_victory ($board) {
   say "\nGAME OVER";
   my $difference = $board->[PLAYER_HOME] - $board->[COMPUTER_HOME];
   if ($difference < 0) {
      say 'I WIN BY ', -$difference, ' POINTS';
   }
   else {
      say $difference ? "YOU WIN BY $difference POINTS" : 'DRAWN GAME';
   }
   return $difference;
}

# move the seeds from $pit and take into account possible bonuses
sub move_seeds ($board, $pit) {

   # get the seeds from the selected pit $pit
   my $seeds = $board->[$pit];
   $board->[$pit] = 0;

   # $landing will be our "moving cursor" to place seeds around
   my $landing = $pit;
   while ($seeds > 0) {
      $landing = ($landing + 1) % 14; # 12 --> 13 -[wrap]-> 0 --> 1
      --$seeds;
      ++$board->[$landing];
   }

   # check for "stealing seeds" condition. This cannot happen in home pits
   if ($landing != PLAYER_HOME && $landing != COMPUTER_HOME
       && $board->[$landing] == 1 && $board->[12 - $landing] > 0) {
      my $home = $pit < 7 ? PLAYER_HOME : COMPUTER_HOME;
      $board->[$home] += 1 + $board->[12 - $landing];
      $board->@[$landing, 12 - $landing] = (0, 0);
   }

   return ($pit, $landing);
}

sub get_player_move ($board, $prompt) {
   print "\n$prompt? ";
   while (defined(my $move = <STDIN>)) {
      chomp($move); # remove newline
      return $move - 1 if $move =~ m{\A[1-6]\z}mxs && $board->[$move - 1];
      print 'ILLEGAL MOVE\nAGAIN? ';
   }
   die "goodbye\n";
}

sub player_move ($board, $stage = FIRST) {
   my $prompt = $stage == FIRST ? 'YOUR MOVE' : 'AGAIN';
   my $selected_move = get_player_move($board, $prompt);
   return move_seeds($board, $selected_move);
}

sub computer_move ($board, $failures, $moves) {

   # we will go through all possible moves for the computer and all
   # possible responses by the player, collecting the "best" move in terms
   # of reasonable outcome (assuming that each side wants to maximize their
   # outcome. $best_move will eventually contain the best move for the
   # computer, and $best_difference the best difference in scoring (as
   # seen from the computer).
   my ($best_move, $best_difference);
   for my $c_move (7 .. 12) {
      next unless $board->[$c_move]; # only consider pits with seeds inside

      # we work on a copy of the board to do all our trial-and-errors
      my $copy = [ $board->@* ];
      move_seeds($copy, $c_move);

      # it's time to "think like a player" and see what's the "best" move
      # for the player in this situation. This heuristic is "not perfect"
      # but it seems OK anyway.
      my $best_player_score = 0;
      for my $p_move (0 .. 5) {
         next unless $copy->[$p_move]; # only pits with seeds inside
         my $landing = $copy->[$p_move] + $p_move;

         # the player's score for this move, calculated as additional seeds
         # placed in the player's pit. The original algorithm sets this to
         # 1 only if the $landing position is greater than 13, which can
         # be obtained by setting the ORIGINAL environment variable to a
         # "true" value (in Perl terms). Otherwise it is calculated
         # according to the real rules for the game.
         my $p_score = $ENV{ORIGINAL} ? $landing > 13 : int(($landing - 5) / 14);

         # whatever, the landing position must be within the bounds
         $landing %= 14;

         # if the conditions apply, the player's move might win additional
         # seeds, which we have to to take into account.
         $p_score += $copy->[12 - $landing]
            if $copy->[$landing] == 0
            && $landing != PLAYER_HOME && $landing != COMPUTER_HOME;

         # let's compare this move's score against the best collected
         # so far (as a response to a specific computer's move).
         $best_player_score = $p_score if $p_score > $best_player_score;
      }

      # the overall score for the player is the additional seeds we just
      # calculated into $best_player_score plus the seeds that were already
      # in the player's pit
      $best_player_score += $copy->[PLAYER_HOME];

      # the best difference we can aim for with this computer's move must
      # assume that the player will try its best
      my $difference = $copy->[COMPUTER_HOME] - $best_player_score;

      # now it's time to check this computer's move against the history
      # of failed matches. $candidate_moves will be the "candidate" list
      # of moves if we accept this one.
      my $candidate_moves = $moves . $c_move . '/';
      for my $failure ($failures->@*) {

         # index(.) returns 0 if and only if $candidate_moves appears at
         # the very beginning of $failure, i.e. it matches a previous
         # behaviour.
         next if index($failure, $candidate_moves) != 0;

         # same sequence of moves as before... assign a penalty
         $difference -= 2;
      }

      # update $best_move and $best_difference if they need to
      ($best_move, $best_difference) = ($c_move, $difference)
         if (! defined $best_move) || ($best_difference < $difference);
   }

   # apply the selected move and return
   return move_seeds($board, $best_move);
}

sub welcome {
   say ' ' x 34, 'AWARI';
   say ' ' x 15, 'CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY';
}

sub print_board ($board) {
   my $template = '
    %2d  %2d  %2d  %2d  %2d  %2d
 %2d                         %d
    %2d  %2d  %2d  %2d  %2d  %2d
';
   printf $template, $board->@[12, 11, 10, 9, 8 , 7, 13, 6, 0 .. 5];
   return;
}

sub is_game_over ($board) {

   # game over if the player's side is empty
   return 1 if none { $_ } $board->@[0 ..  5];

   # game over if the computers' side is empty
   return 1 if none { $_ } $board->@[7 .. 12];

   # not game over
   return 0;
}

File: 04_Awari/ruby/awari.rb
require 'strscan'

def print_beans(beans)
  puts

  # Print computer beans
  print "   "
  beans[7..12].reverse.each {|bean_count| print_bean(bean_count)}
  puts

  # Print home beans
  print_bean(beans[13])
  print '                       '
  print_number(beans[6]) # This is not print_bean in line with the original version
  puts

  # Print player beans
  print "   "
  beans[0..5].each {|bean_count| print_bean(bean_count)}
  puts

  puts
end

def print_bean(bean_count)
  print ' ' if bean_count < 10
  print_number(bean_count)
end

def print_number(n)
  # PRINT adds padding after a number and before a positive number
  print ' ' if n >= 0
  print n.to_s
  print ' '
end

def get_move(prompt, beans)
  move = get_integer_input(prompt)

  while move < 1 || move > 6 || beans[move - 1] == 0
    puts "ILLEGAL MOVE"
    move = get_integer_input("AGAIN")
  end

  move - 1
end

def get_integer_input(prompt)
  integer_value = nil

  input_values = input(prompt)

  while integer_value.nil?
    print '!EXTRA INPUT IGNORED' if (input_values.size > 1)

    value = input_values.first

    begin
      integer_value = Integer(value)
    rescue
      puts '!NUMBER EXPECTED - RETRY INPUT LINE'
      input_values = input('')
    end
  end

  integer_value
end

def input(prompt)
  prompt_suffix = '? '
  print "#{prompt}#{prompt_suffix}"

  input = gets.chomp.strip
  scanner = StringScanner.new(input)
  input_values = []

  until scanner.eos?
    scanner.scan(/\s+/)

    if scanner.check(/"/)
      scanner.scan(/"/)
      next_string = scanner.scan_until(/"/)

      if next_string
        # Remove the trailing close quote
        next_string.chomp!('"')
      else
        # No close quote – Vintage Basic crashes in this case
        raise ArgumentError('Unmatched quotes in input')
      end
    elsif scanner.exist?(/,/)
      next_string = scanner.scan_until(/,/).chomp(',')
    else
      next_string = scanner.scan_until(/\s+|$/).rstrip
    end

    input_values << next_string
  end

  input_values << '' if input_values.empty?

  input_values
end


def distribute_beans(beans, start_pit, home_pit)
  beans_to_distribute = beans[start_pit]
  beans[start_pit] = 0

  current_pit = start_pit

  (0...beans_to_distribute).each do
    current_pit = (current_pit + 1) % beans.size
    beans[current_pit] += 1
  end

  # If the last pit was empty before we put a bean in it (and it's not a scoring pit), add beans to score
  if beans[current_pit] == 1 && current_pit != 6 && current_pit != 13 && beans[12 - current_pit] != 0
    beans[home_pit] = beans[home_pit] + beans[12 - current_pit] + 1
    beans[current_pit] = 0
    beans[12 - current_pit] = 0
  end

  current_pit
end

def update_internals(beans, current_move)
  k = current_move % 7
  $c = $c + 1

  $f[$n] = $f[$n] * 6 + k if $c < 9

  unless beans[0...6].find { |b| b != 0 } && beans[7...13].find { |b| b != 0 }
    $game_over = true
  end
end

# @param [Array] beans
# @param [Integer] move
# @param [Integer] home_pit
def perform_move(beans, move, home_pit)
  last_pit = distribute_beans(beans, move, home_pit)

  update_internals(beans, move)

  last_pit
end

def end_game(beans)
  puts
  puts "GAME OVER"

  difference = beans[6] - beans[13]

  if difference < 0
    puts "I WIN BY #{-difference} POINTS"

    return
  end

  $n += 1

  puts "YOU WIN BY #{difference} POINTS" if difference > 0
  puts "DRAWN GAME" if difference == 0
end

# @param [Array] beans
def get_computer_move(beans)
  d = -99
  home_pit = 13

  chosen_move = 7

  # Test all possible moves
  (7...13).each do |move_under_test|
    # Create a copy of the beans to test against
    beans_copy = beans.dup

    # If the move is not legal, skip it
    next if beans[move_under_test] == 0

    # Determine the best response the player may make to this move
    player_max_score = 0

    # Make the move under test against the copy
    distribute_beans(beans_copy, move_under_test, home_pit)

    # Test every player response
    (0...6).each do |i|
      # Skip the move if it would be illegal
      next if beans_copy[i] == 0

      # Determine the last
      landing_with_overflow = beans_copy[i] + i
      # If landing > 13 the player has put a bean in both home pits
      player_move_score = (landing_with_overflow > 14) ? 1 : 0
      # Find the actual pit
      landing = landing_with_overflow % 14

      # If the landing pit is empty, the player will steal beans
      if beans_copy[landing] == 0 && landing != 6 && landing != 13
        player_move_score = beans_copy[12 - landing] + player_move_score
      end

      # Update the max score if this move is the best player move
      player_max_score = player_move_score if player_move_score > player_max_score
    end

    # Final score for move is computer score, minus the player's score and any player gains from their best move
    final_score = beans_copy[13] - beans_copy[6] - player_max_score

    if $c <=8
      k = move_under_test % 7

      (0...$n).each do |i|
        final_score = final_score - 2 if $f[$n] * 6 + k == ((Float($f[i])/6 ** (7-$c)) + 0.1).floor
      end
    end

    # Choose the move if it is the best move found so far
    if final_score >= d
      chosen_move = move_under_test
      d = final_score
    end
  end

  last_pit = perform_move(beans, chosen_move, home_pit)

  [chosen_move, last_pit]
end

puts 'AWARI'.center(80)
puts 'CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY'.center(80)

# Initialise stable variables
$f = Array.new(50)
$n = 0

# APPLICATION LOOP
:game
while true
  print "\n", "\n"
  beans = Array.new(13, 3)
  beans[6] = 0
  beans[13] = 0

  $f[$n] = 0
  $game_over = false
  $c = 0

  until $game_over
    print_beans(beans)

    move = get_move("YOUR MOVE", beans)
    home_pit = 6
    computer_home_pit = 13

    last_pit = perform_move(beans, move, home_pit)

    print_beans(beans)

    if $game_over
      end_game(beans)
      next :game
    end

    if home_pit == last_pit
      second_move = get_move("AGAIN", beans)

      perform_move(beans, second_move, home_pit)

      print_beans(beans)

      if $game_over
        end_game(beans)
        next :game
      end
    end

    computer_move, computer_last_pit = get_computer_move(beans)
    print "MY MOVE IS #{computer_move - 6}"

    if $game_over
      end_game(beans)
      next :game
    end

    if computer_last_pit == computer_home_pit
      second_computer_move, _ = get_computer_move(beans)
      print ",#{second_computer_move - 6}"

      if $game_over
        end_game(beans)
        next :game
      end
    end
  end
end
